File: sdk/typescript/src/session.ts
/**
 * Action lifecycle and idempotency store (SPEC §6).
 */
import type { ResultStatus } from "./messages.js";

export const RESULT_TO_STATE: Record<ResultStatus, string> = {
  accepted: "ACCEPTED",
  running: "RUNNING",
  ok: "SUCCESS",
  failed: "FAILED",
  timeout: "TIMEOUT",
  rejected: "REJECTED",
};

const TERMINAL = new Set(["SUCCESS", "FAILED", "TIMEOUT", "REJECTED"]);
const TERMINAL_STATUS: ResultStatus[] = ["ok", "failed", "timeout", "rejected"];

export type MarkResult = "ok" | "invalid_status" | "unknown_action" | "after_terminal";

export class ActionStore {
  actions = new Map<string, { state: string; name: string; source: string }>();
  outcomes = new Map<string, [ResultStatus, string | null]>();
  violations: Array<[string, string, string]> = [];

  register(actionId: string, name: string, source: string): void {
    if (!this.actions.has(actionId)) {
      this.actions.set(actionId, { state: "PENDING", name, source });
    }
  }

  mark(actionId: string, status: ResultStatus, resultId?: string): MarkResult {
    const state = RESULT_TO_STATE[status];
    if (!state) return "invalid_status";
    const action = this.actions.get(actionId);
    if (!action) return "unknown_action";
    if (TERMINAL.has(action.state) && !TERMINAL_STATUS.includes(status)) {
      this.violations.push(["I6", actionId, status]);
      return "after_terminal";
    }
    if (status === "accepted" && action.state === "PENDING") {
      action.state = "ACCEPTED";
    } else if (status === "running") {
      action.state = "RUNNING";
    } else {
      action.state = state;
    }
    if (TERMINAL_STATUS.includes(status)) {
      this.outcomes.set(actionId, [status, resultId ?? null]);
    }
    return "ok";
  }

  state(actionId: string): string | undefined {
    return this.actions.get(actionId)?.state;
  }

  outcome(actionId: string): [ResultStatus, string | null] | undefined {
    return this.outcomes.get(actionId);
  }
}

export class Session {
  sessionId: string;
  state: "ACTIVE" | "EXPIRED" = "ACTIVE";
  actions = new ActionStore();
  receiver: import("./sequence.js").Receiver | null = null;

  constructor(sessionId: string) {
    this.sessionId = sessionId;
  }

  cursors(): Record<string, number> {
    return this.receiver ? this.receiver.cursors(this.sessionId) : {};
  }

  expire(): void {
    this.state = "EXPIRED";
  }

  isActive(): boolean {
    return this.state === "ACTIVE";
  }
}

File: sdk/python/aip/session.py
"""Action lifecycle and idempotency store (SPEC §6)."""
from typing import Optional, Tuple

RESULT_TO_STATE = {
    "accepted": "ACCEPTED",
    "running": "RUNNING",
    "ok": "SUCCESS",
    "failed": "FAILED",
    "timeout": "TIMEOUT",
    "rejected": "REJECTED",
}
TERMINAL = {"SUCCESS", "FAILED", "TIMEOUT", "REJECTED"}
TERMINAL_STATUS = ("ok", "failed", "timeout", "rejected")


class ActionStore:
    """Tracks per-action lifecycle state and idempotency outcomes (SPEC §6)."""

    def __init__(self) -> None:
        self.actions: dict = {}   # action_id -> {state, name, source}
        self.outcomes: dict = {}  # action_id -> (status, result_id) once terminal
        self.violations: list = []  # (invariant, action_id, detail)

    def register(self, action_id: str, name: str, source: str) -> dict:
        if action_id not in self.actions:
            self.actions[action_id] = {"state": "PENDING", "name": name, "source": source}
        return self.actions[action_id]

    def mark(self, action_id: str, status: str, result_id: Optional[str] = None) -> str:
        """Apply a result status to an action.

        Returns "ok" normally, or one of:
        "invalid_status", "unknown_action", "after_terminal" (I6 violation).
        """
        state = RESULT_TO_STATE.get(status)
        if state is None:
            return "invalid_status"
        action = self.actions.get(action_id)
        if action is None:
            return "unknown_action"
        if action["state"] in TERMINAL and status not in TERMINAL_STATUS:
            self.violations.append(("I6", action_id, status))
            return "after_terminal"
        if status == "accepted" and action["state"] == "PENDING":
            action["state"] = "ACCEPTED"
        elif status == "running":
            action["state"] = "RUNNING"
        else:
            action["state"] = state
        if status in TERMINAL_STATUS:
            self.outcomes[action_id] = (status, result_id)
        return "ok"

    def state(self, action_id: str) -> Optional[str]:
        action = self.actions.get(action_id)
        return action["state"] if action else None

    def outcome(self, action_id: str) -> Optional[Tuple[str, Optional[str]]]:
        return self.outcomes.get(action_id)


class Session:
    """A session is externalized state: cursors, actions, lifecycle (SPEC §7.2)."""

    def __init__(self, session_id: str) -> None:
        self.session_id = session_id
        self.state = "ACTIVE"
        self.actions = ActionStore()
        self.receiver = None  # injected by the peer/gateway that owns the wire

    def cursors(self) -> dict:
        return self.receiver.cursors(self.session_id) if self.receiver else {}

    def expire(self) -> None:
        self.state = "EXPIRED"

    def is_active(self) -> bool:
        return self.state == "ACTIVE"